stingray/pulse/pulsar.py

```diff
@@ -313,31 +414,201 @@ def z_n(phase, n=2, norm=1):
     n : int, default 2
         Number of harmonics, including the fundamental
 
+    Returns
+    -------
+    ks : list of ints
+        Harmonic numbers, from 1 to nmax (included)
+    z2_n : float
+        The Z^2_n statistic for all ks
+    '''
+    all_zs = np.zeros(nmax)
+    ks = np.arange(1, nmax + 1)
+    nphot = phase.size
+
+    total_sum = 0
+    phase = phase * 2 * np.pi
+
+    for k in ks:
+        local_z = (
+            np.sum(np.cos(k * phase)) ** 2
+            + np.sum(np.sin(k * phase)) ** 2
+        )
+        total_sum += local_z
+        all_zs[k - 1] = total_sum
+
+    return ks, 2 / nphot * all_zs
+
+
+def z_n_poisson(profile, n):
+    '''Z^2_n statistic for Poissonian pulse profiles
+
+    See Bachetti+2021, arXiv:2012.11397
+
+    Parameters
+    ----------
+    profile : array of floats
+        The folded pulse profile (containing the number of
+        photons falling in each pulse bin)
+    n : int
+        Number of harmonics, including the fundamental
+
+    Returns
+    -------
+    z2_n : float
+        The value of the statistic
+    '''
+    _, all_zs = z_n_poisson_all(profile, nmax=n)
+    return all_zs[-1]
```

Comment:
I am confused why this is a separate function, if all it does is call `z_n_poisson_all`? Could this be merged into a single function, or am I missing something? Same with the other versions below.

Reply:
Basically, one needs to calculate all k (1..n) values to get the Z^2_n. For the H test, we need all n values, while for the Z statistic we need only the last one. At the expense of some memory usage to store all n values and only use the last one, I obtained a lot less code duplication